3/assets/3.6-finished-code.rs

```diff
@@ -0,0 +1,315 @@
+            ensure!(<Kitties<T>>::exists(kitty_id_2), "This cat 2 does not exist");
+
+            let nonce = <Nonce<T>>::get();
+            let random_hash = (<system::Module<T>>::random_seed(), &sender, nonce)
+                .using_encoded(<T as system::Trait>::Hashing::hash);
+
+            let kitty_1 = Self::kitty(kitty_id_1);
+            let kitty_2 = Self::kitty(kitty_id_2);
+
+            let mut final_dna = kitty_1.dna;
+
+            for (i, (dna_2_element, r)) in kitty_2.dna.as_ref().iter().zip(random_hash.as_ref().iter()).enumerate() {
+                if r % 2 == 0 {
+                    final_dna.as_mut()[i] = *dna_2_element;
+                }
+            }
+
+            let new_kitty = Kitty {
+                id: random_hash,
+                dna: final_dna,
+                price: <T::Balance as As<u64>>::sa(0),
+                gen: cmp::max(kitty_1.gen, kitty_2.gen) + 1,
+            };
+
+            Self::_mint(sender, random_hash, new_kitty)?;
+
+            <Nonce<T>>::mutate(|n| *n += 1);
+
+            Ok(())
+        }
+
+        fn create_auction(origin, kitty_id: T::Hash, min_bid: T::Balance, expiry_in_minutes: u64) -> Result {
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+            let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+            ensure!(owner == sender, "You can't set an auction for a cat you don't own");
+
+            let expiry = <system::Module<T>>::block_number() + T::BlockNumber::sa(expiry_in_minutes*12);
+            ensure!(expiry <= <system::Module<T>>::block_number() + Self::auction_duration_limit(), "The auction duration should be less than 5 days");
+
+            let new_auction = Auction {
+                kitty_id,
+                kitty_owner: owner,
+                expiry,
+                min_bid,
+                high_bid: min_bid,
+                high_bidder: sender,
+            };
+
+            <KittyAuction<T>>::insert(kitty_id, &new_auction);
+            <Auctions<T>>::insert(expiry, vec![new_auction.clone()]);
+
+            Self::deposit_event(RawEvent::Auction(kitty_id, min_bid, expiry));
+
+            Ok (())
+        }
+
+        fn bid_auction(origin, kitty_id: T::Hash, bid: T::Balance) -> Result {
+            let sender = ensure_signed(origin)?;
+
+            ensure!(<Kitties<T>>::exists(kitty_id), "This cat does not exist");
+
+            let owner = Self::owner_of(kitty_id).ok_or("No owner for this kitty")?;
+            ensure!(owner != sender, "You can't bid for your own cat");
+
+            let mut auction = Self::auction_of(kitty_id).ok_or("No auction for this cat")?;
+            ensure!(<system::Module<T>>::block_number() < auction.expiry, "This auction is expired.");
+
+            ensure!(bid > auction.min_bid, "Your bid has to be greater than the minimum bid.");
+
```

Comment:
You have at no point checked that the user actually has enough funds for this bid.

You need to be a bit more fancy here. You need to:
* Check the user has enough funds to make a bid
* Escrow the user's bid, and track how much you took from them
* Make sure new bids from the same user take into account how much we have already escrowed
* Return all funds to losing users at the end of the bidding war

Reply:
This will probably require more storage items

Reply:
Just to make sure, since a module does not have a balance (as opposed to a smart contract), I will let the owner add an `escrow_account: T::AccountId` in `create_auction()`.

Then, transfer bid amounts from bidders to this account in `bid_auction()` and keep a record of it in the storage. 

While resolving the auction in `on_finalise()`, transfer the losing bidder their amounts from the escrow account, and transfer the winning bidder's bid from escrow to the kitty owner. 

Is this a right way to go? However, this is creating an arbitrary account as an escrow, would there be another way with less trust?

Reply:
I think you want to use the `ReservedBalance`:
https://github.com/paritytech/substrate/blob/master/srml/balances/src/lib.rs#L148

This balance "stays" with the user, but they dont really have access to it (they cannot transfer it). You need to track how much you put there, and then put it back into `free_balance` once you are done. (Putting back into free balance should actually be one of the checks that need to pass before you actually complete everything)